rfc/custom-modules-v1.md

```diff
@@ -0,0 +1,706 @@
+---
+title: "Router Custom Module System v1"
+author: Dustin Deus
+date: 2024-08-11
+status: Draft
+---
+
+# Router Custom Module System v1
+
+- **Author:** Dustin Deus
+- **Date:** 2024-08-11
+- **Status:** Draft
+
+## Abstract
+
+This RFC describes an overhaul of the current module system in the router. The new module system is designed to be more flexible and native to GraphQL. It allows developers to hook into the router lifecycle as well as outgoing and incoming requests to subgraphs.
+
+## Introduction
+
+As of today, customers can extend the router with custom modules. These modules can be used to change the behavior of the router, add custom logic, or integrate with other systems. The current module system has several limitations that we want to address with this RFC:
+
+- The current module system is not native to GraphQL. It is based on HTTP middleware and does not provide a GraphQL-specific API.
+- The current module system is inconsistent and hard to use. It does not provide a clear API for developers to intercept and modify GraphQL requests and responses.
```

Comment:
Hi,

Great stuff here. One thing that I didn't see is the possibility to hook into the EDFS lifecycle.
Would it be possible with the Custom Module System to intercept subscription requests to subgraphs and inject a custom transport, similar to the Kafka or NATS  that you already provide.

I would be particularly interested in adding Redis PubSub support to EDFS.

Thx,
Goran

Reply:
Hi @ghorak-happening, that's a great idea. The goal of this RFC is to propose an architecture that is flexible enough to add those extensions later. The first version will focus on the fundamentals. I'm going to add your requirements to the list so as not to forget them.

Reply:
That would be great @StarpTech, thank you!